Assets/MixedRealityToolkit.Services/InputSimulation/KeyBinding.cs

```diff
@@ -0,0 +1,255 @@
+        /// The type of value encoded in the <see cref="code"/> property.
+        /// </summary>
+        public enum KeyType : int
+        {
+            None = 0,
+            Mouse = 1,
+            Key = 2,
+        }
+
+        /// <summary>
+        /// Enum for interpreting the mouse button integer index.
+        /// </summary>
+        public enum MouseButton : int
+        {
+            Left = 0,
+            Right = 1,
+            Middle = 2,
+            Button3 = 3,
+            Button4 = 4,
+            Button5 = 5,
+            Button6 = 6,
+            Button7 = 7,
+        }
+
+        // Array of names to use for a combined enum selection.
+        internal static readonly string[] AllCodeNames;
+        // Maps (KeyType, code) combination onto the contiguous index used for enums.
+        // Value can be used as index in AllCodeNames array.
+        internal static readonly Dictionary<Tuple<KeyType, int>, int> KeyBindingToEnumMap;
+        // Maps enum index to a KeyBinding, for assignment after selecting an enum value.
+        internal static readonly Dictionary<int, KeyBinding> EnumToKeyBindingMap;
+
+        // Static constructor to initialize static fields
+        static KeyBinding()
+        {
+            KeyCode[] KeyCodeValues = (KeyCode[])Enum.GetValues(typeof(KeyCode));
+            string[] KeyCodeNames = Enum.GetNames(typeof(KeyCode));
+            MouseButton[] MouseButtonValues = (MouseButton[])Enum.GetValues(typeof(MouseButton));
+            string[] MouseButtonNames = Enum.GetNames(typeof(MouseButton));
+
+            // Build maps for converting between int enum value and KeyBinding values
+            {
+                KeyBindingToEnumMap = new Dictionary<Tuple<KeyType, int>, int>();
+                EnumToKeyBindingMap = new Dictionary<int, KeyBinding>();
+                List<string> names = new List<string>();
+
+                int index = 0;
+                Action<KeyType, int> AddEnumValue = (bindingType, code) =>
+                {
+                    var kb = new KeyBinding() { bindingType=bindingType, code=code };
+                    names.Add(kb.ToString());
+                    EnumToKeyBindingMap[index] = kb;
+                    KeyBindingToEnumMap[Tuple.Create(bindingType, code)] = index;
+
+                    ++index;
+                };
+
+                AddEnumValue(KeyType.None, 0);
+
+                foreach (MouseButton mb in MouseButtonValues)
+                {
+                    AddEnumValue(KeyType.Mouse, (int)mb);
+                }
+
+                foreach (KeyCode kc in KeyCodeValues)
+                {
+                    AddEnumValue(KeyType.Key, (int)kc);
+                }
+
+                AllCodeNames = names.ToArray();
+            }
+        }
+
+        [SerializeField]
+        private KeyType bindingType;
+        public KeyType BindingType => bindingType;
+        [SerializeField]
+        private int code;
+        internal int Code => code;
```

Comment:
>internal [](start = 8, length = 8)

Was this supposed to be public? I haven't seen this private/internal combination in the codebase yet.

Reply:
These are marked internal so that the inspector can access them. I didn't want to make them public because the value is not very meaningful on its own, the TryGetKeyCode and TryGetMouseButton functions should be used instead if one needs to know which internal KeyCode or mouse button is used.

The inspector needs access to some internals of the KeyBinding to convert to and from a plain enum index, which allows selecting a KeyBinding directly from a combined list. I've changed the inspector slightly so it now accesses the enum dictionaries and doesn't need the Code property.